Web Inspector: include `referrerPolicy` in "Copy as fetch"

dcrousso opened the pull request:
#### 531afa8f61ee4bb09cc15dea91bd0dc280ef0fd1
<pre>
Web Inspector: include `referrerPolicy` in &quot;Copy as fetch&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=241218">https://bugs.webkit.org/show_bug.cgi?id=241218</a>
&lt;rdar://problem/94698964&gt;

Reviewed by Patrick Angle.

* Source/JavaScriptCore/inspector/protocol/Network.json:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::toProtocol):
(WebCore::buildObjectForResourceRequest):
(WebCore::InspectorNetworkAgent::willSendRequest):
(WebCore::InspectorNetworkAgent::willSendRequestOfType):
(WebCore::InspectorNetworkAgent::interceptRequest):
Add a `Network.ReferrerPolicy` protocol enum (corresponding to `WebCore::ReferrerPolicy`) and pass it to the frontend.

* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::willSendRequest):
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::willSendRequestImpl):
* Source/WebCore/loader/ResourceLoadNotifier.h:
* Source/WebCore/loader/ResourceLoadNotifier.cpp:
(WebCore::ResourceLoadNotifier::willSendRequest):
(WebCore::ResourceLoadNotifier::dispatchWillSendRequest):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::willSendRequestInternal):
* Source/WebCore/loader/appcache/ApplicationCacheGroup.cpp:
(WebCore::ApplicationCacheGroup::update):
(WebCore::ApplicationCacheGroup::startLoadingEntry):
Pass along a `ResourceLoader*` when `willSendRequest` so that the `InspectorNetworkAgent` can look
at it to get the `FetchOptions` for that request.

* Source/WebInspectorUI/UserInterface/Models/Resource.js:
(WI.Resource.prototype.get referrerPolicy): Added.
(WI.Resource.prototype.updateForRedirectResponse):
(WI.Resource.prototype.generateFetchCode):
* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.prototype.resourceRequestWillBeSent):
Pass along the `Network.ReferrerPolicy` whenever creating/updating a `WI.Resource` with a `Network.Request`.

* Source/WebCore/inspector/agents/InspectorPageAgent.h:
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::cachedResource):
Drive-by: Change to `const Frame*`.

* LayoutTests/http/tests/inspector/network/copy-as-fetch.html:
* LayoutTests/http/tests/inspector/network/copy-as-fetch-expected.txt:

Canonical link: <a href="https://commits.webkit.org/251818@main">https://commits.webkit.org/251818@main</a>
</pre>

patrickangle commented on Source/JavaScriptCore/inspector/protocol/Network.json:
```diff
@@ -76,6 +91,7 @@
                 { "name": "mimeType", "type": "string", "description": "Resource mimeType as determined by the browser." },
                 { "name": "source", "type": "string", "enum": ["unknown", "network", "memory-cache", "disk-cache", "service-worker", "inspector-override"], "description": "Specifies where the response came from." },
                 { "name": "requestHeaders", "$ref": "Headers", "optional": true, "description": "Refined HTTP request headers that were actually transmitted over the network." },
+                { "name": "referrerPolicy", "$ref": "ReferrerPolicy", "optional": true },
```
No description?

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/Resource.js:
```diff
@@ -1200,7 +1207,9 @@ WI.Resource = class Resource extends WI.SourceCode
-        // FIXME: <https://webkit.org/b/241218> Web Inspector: include `referrerPolicy` in "Copy as fetch"
+        // COMPATIBILITY (iOS 15.4): Network.ReferrerPolicy did not exist yet.
+        if (this._referrerPolicy)
+            options.referrerPolicy = this._referrerPolicy;
```
Not sure this warrants a `COMPATIBILITY` comment here since even post-iOS 15.4 an empty string will be falsey, and I assume it is intentional that we would then skip adding it to the created code (the test makes it seem like it is intentional).

dcrousso replied:
my thinking was that we should always have a `referrerPolicy` for every request (i.e. we'd eventually be able to remove this `if`), but I guess the protocol has it as optional so yeah it makes sense to also check it here

patrickangle commented on Source/JavaScriptCore/inspector/protocol/Network.json:
```diff
@@ -29,6 +29,21 @@
+        {
+            "id": "ReferrerPolicy",
+            "type": "string",
+            "enum": [
```
No description?

dcrousso committed:
Web Inspector: include `referrerPolicy` in "Copy as fetch"
https://bugs.webkit.org/show_bug.cgi?id=241218
<rdar://problem/94698964>

Reviewed by Patrick Angle.

* Source/JavaScriptCore/inspector/protocol/Network.json:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::toProtocol):
(WebCore::buildObjectForResourceRequest):
(WebCore::InspectorNetworkAgent::willSendRequest):
(WebCore::InspectorNetworkAgent::willSendRequestOfType):
(WebCore::InspectorNetworkAgent::interceptRequest):
Add a `Network.ReferrerPolicy` protocol enum (corresponding to `WebCore::ReferrerPolicy`) and pass it to the frontend.

* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::willSendRequest):
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::willSendRequestImpl):
* Source/WebCore/loader/ResourceLoadNotifier.h:
* Source/WebCore/loader/ResourceLoadNotifier.cpp:
(WebCore::ResourceLoadNotifier::willSendRequest):
(WebCore::ResourceLoadNotifier::dispatchWillSendRequest):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::willSendRequestInternal):
* Source/WebCore/loader/appcache/ApplicationCacheGroup.cpp:
(WebCore::ApplicationCacheGroup::update):
(WebCore::ApplicationCacheGroup::startLoadingEntry):
Pass along a `ResourceLoader*` when `willSendRequest` so that the `InspectorNetworkAgent` can look
at it to get the `FetchOptions` for that request.

* Source/WebInspectorUI/UserInterface/Models/Resource.js:
(WI.Resource.prototype.get referrerPolicy): Added.
(WI.Resource.prototype.updateForRedirectResponse):
(WI.Resource.prototype.generateFetchCode):
* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.prototype.resourceRequestWillBeSent):
Pass along the `Network.ReferrerPolicy` whenever creating/updating a `WI.Resource` with a `Network.Request`.

* Source/WebCore/inspector/agents/InspectorPageAgent.h:
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::cachedResource):
Drive-by: Change to `const Frame*`.

* LayoutTests/http/tests/inspector/network/copy-as-fetch.html:
* LayoutTests/http/tests/inspector/network/copy-as-fetch-expected.txt:

Canonical link: https://commits.webkit.org/251818@main
